core/src/main/java/io/grpc/util/GracefulSwitchLoadBalancer.java

```diff
@@ -0,0 +1,158 @@
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.ConnectivityState;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.Internal;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.Status;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A load balancer that gracefully swaps to a new lb policy. If the channel is currently in a state
+ * other than READY, the new policy will be swapped into place immediately.  Otherwise, the channel
+ * will keep using the old policy until the new policy reports READY or the old policy exits READY.
+ */
+@Internal
+@NotThreadSafe // Must be accessed in SynchronizationContext
+public final class GracefulSwitchLoadBalancer extends ForwardingLoadBalancer {
+  private static final LoadBalancer NOOP_BALANCER = new LoadBalancer() {
+    @Override
+    public void handleNameResolutionError(Status error) {}
+
+    @Override
+    public void shutdown() {}
+  };
+
+  @VisibleForTesting
+  static final SubchannelPicker BUFFER_PICKER = new SubchannelPicker() {
+    @Override
+    public PickResult pickSubchannel(PickSubchannelArgs args) {
+      return PickResult.withNoResult();
+    }
+
+    @Override
+    public String toString() {
+      return "BUFFER_PICKER";
+    }
+  };
+
+  private final Helper helper;
+  private LoadBalancer currentLb = NOOP_BALANCER;
+  private LoadBalancer pendingLb = NOOP_BALANCER;
+  @Nullable
+  private String currentPolicyName;
+  @Nullable
+  private String pendingPolicyName;
+  private boolean isReady;
+  private ConnectivityState pendingState;
+  private SubchannelPicker pendingPicker;
+
+  public GracefulSwitchLoadBalancer(Helper helper) {
+    this.helper = checkNotNull(helper, "helper");
+  }
+
+  /** Gracefully switch to a new load balancing policy. */
+  public void switchTo(LoadBalancerProvider newLbProvider) {
+    checkNotNull(newLbProvider, "newLbProvider");
+
+    String newPolicyName = newLbProvider.getPolicyName();
+    if (newPolicyName.equals(pendingPolicyName)) {
+      return;
+    }
+    pendingLb.shutdown();
+    pendingLb = NOOP_BALANCER;
+    pendingPolicyName = null;
+    pendingState = ConnectivityState.IDLE;
```

Comment:
I thought it was created with IDLE. `ManagedChannelImpl.exitIdleMode()` doesn't change the state from IDLE to CONNECTING.

Reply:
Hmmm... you're right, it doesn't. That seems like a bug. The initial name resolution should be considered part of CONNECTING. And when we create a LB we do expect it to connect to addresses as soon as it gets the list, without any RPC being issued.

Maybe that's what this comment is referring to (although the `updateBalancingState()` seems unnecessary, because the SubchannelStateListener is going to fire immediately and we'll update the balancer state again):
https://github.com/grpc/grpc-java/blob/8a9afd618aec0e1bf5858232cac70c89a90fb96e/core/src/main/java/io/grpc/internal/PickFirstLoadBalancer.java#L62-L64

Reply:
>although the updateBalancingState() seems unnecessary, because the SubchannelStateListener is going to fire immediately and we'll update the balancer state again

So because of above, can we set initial pendingState = null, and if old policy goes out of READY state while pendingState is null, we don't update balancing state immediately, it will be updated automatically on the new policy's inital handleResolvedAddress()?

Reply:
If we don't update the state, that implies we are still using the old policy. We should not continue using the picker for a shutdown() policy. So I do think it is important to make sure we replace the picker. I made a small comment to make that clearer in swap.

Reply:
Done.